x/cash/dynamicfee.go

```diff
@@ -42,98 +42,63 @@ import (
 
 type DynamicFeeDecorator struct {
 	auth x.Authenticator
-	ctrl CoinMover
+	ctrl Controller
 }
 
 var _ weave.Decorator = DynamicFeeDecorator{}
 
 // NewDynamicFeeDecorator returns a DynamicFeeDecorator with the given
 // minimum fee, and all collected fees going to a default address.
-func NewDynamicFeeDecorator(auth x.Authenticator, ctrl CoinMover) DynamicFeeDecorator {
+func NewDynamicFeeDecorator(auth x.Authenticator, ctrl Controller) DynamicFeeDecorator {
 	return DynamicFeeDecorator{
 		auth: auth,
 		ctrl: ctrl,
 	}
 }
 
 // Check verifies and deducts fees before calling down the stack
-func (d DynamicFeeDecorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx,
-	next weave.Checker) (weave.CheckResult, error) {
-
-	// prepare fees for subtransaction -- shared in Check and Deliver
-	var res weave.CheckResult
+func (d DynamicFeeDecorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Checker) (res weave.CheckResult, err error) {
 	fee, payer, cache, err := d.prepare(ctx, store, tx)
 	if err != nil {
-		return res, err
+		return res, errors.Wrap(err, "cannot prepare")
 	}
 
-	minFee := gconf.Coin(store, GconfMinimalFee)
-
-	// do subtransaction in a function for easier error handling
-	res, err = func() (weave.CheckResult, error) {
-		// shadow with local variables...
-		err := d.chargeFee(cache, payer, fee)
-		if err != nil {
-			return weave.CheckResult{}, errors.Wrap(err, "cannot charge fee")
+	defer func() {
+		if err == nil {
+			cache.Write()
+			res.GasPayment += toPayment(fee)
+		} else {
+			cache.Discard()
+			_ = d.chargeMinimalFee(store, payer)
 		}
-		res, err := next.Check(ctx, cache, tx)
-		// TODO: check RequiredFee here and return an error if insufficient
-		return res, err
 	}()
 
-	// on error, rollback, then take the minfee from the store
-	if err != nil {
-		cache.Discard()
-		// if this fails, we aborted early above, we can just ignore return value
-		// this is 2 ops, not 1, for errors, but done to optimize the success case to use 1 not 2
-		_ = d.chargeFee(store, payer, minFee)
-		// return error from the transaction, not the possible error from minFee deduction
-		return res, err
-	}
-
-	// if success, we commit and update the importance
-	cache.Write()
-	res.GasPayment += toPayment(fee)
-	return res, err
+	if err := d.chargeFee(cache, payer, fee); err != nil {
```

Comment:
I am very unsure of how all this relates to the returned error. Along with scoping.

I mean, we have `err` as a return value
And this line defines a local err
Then when we return below, defer is executed and gets the `errors.Wrap(...)` error, right?

I mean, on one hand it looks clean, on the other, this is depending quite a bit on the details of scoping. I would rather avoid shadowing if possible. I mean if the return value was eg. `rerr` for return error, then it is clear if we are refering to the function-local err or the return value one. In general, I would limit shadowing and redefinition of variables as it makes it easy for subtle bugs to drop in.

I think this code is correct, but would prefer to remove this crucial shadow that affects the defer() function

Reply:
I was wondering the same thing. Tests are covering those cases and they seem to work as expected.

I do not know how to write it differently. I could define two anonymous functions or use `goto`. I think they all suffer the same issue.

Reply:
I changed the name of the returned values to ferr and fres (final error, final result), so they are not used at all in the normal code, just references for the defer function to use. Exact same logic and control flow you have, I just renamed one variable to avoid shadowing. I would consider adding an anti-shadowing lint rule.. or maybe I just got used to it from the TypeScript linting I was using in another project, where it is not allowed and prevents a number of gotchas.